cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceQueryAdmin.java

```diff
@@ -83,16 +97,18 @@ public boolean apply(NamespaceMeta namespaceMeta) {
    */
   @Override
   public NamespaceMeta get(Id.Namespace namespaceId) throws Exception {
-    NamespaceMeta ns = nsStore.get(namespaceId);
-    if (ns == null) {
-      throw new NamespaceNotFoundException(namespaceId);
-    }
-    Principal principal = authenticationContext.getPrincipal();
-    Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
-    if (!Principal.SYSTEM.equals(principal) && !filter.apply(namespaceId.toEntityId())) {
-      throw new UnauthorizedException(principal, namespaceId.toEntityId());
+    try {
+      return namespaceMetaCache.get(namespaceId.toEntityId());
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof NamespaceNotFoundException) {
+        throw new NamespaceNotFoundException(namespaceId);
```

Comment:
e.getCause() will give a throwable so to throw that we will need to change the API to throw Throwables which will need all caller to handle throwables. We can wrap it in a Exception but I thought wrapping properly will be better idea.